File: CoreData/ZFRCoreData.m
//
//  Created by Francesco Frison on 4/16/14.
//  Copyright (c) 2014 Ziofritz. All rights reserved.
//

#import "ZFRCoreData.h"
#import <CoreData/CoreData.h>


static NSString * const __databaseName = @"Database.sqlite";
static NSString * const __databaseNameTest = @"TestDatabase.sqlite";
static NSString * const __managedObjectModelName = @"ModelName";

@implementation ZFRCoreData


@synthesize managedObjectContext = _managedObjectContext;
@synthesize managedObjectModel = _managedObjectModel;
@synthesize persistentStoreCoordinator = _persistentStoreCoordinator;

#pragma mark --
#pragma mark - Core Data stack

/**
 Returns the managed object context for the application.
 */
- (NSManagedObjectContext *)managedObjectContext
{
    if (_managedObjectContext != nil)
    {
        return _managedObjectContext;
    }
    
    NSPersistentStoreCoordinator *coordinator = [self persistentStoreCoordinator];
    if (coordinator != nil)
    {
        _managedObjectContext = [[NSManagedObjectContext alloc] init];
        [_managedObjectContext setPersistentStoreCoordinator:coordinator];
    }
    return _managedObjectContext;
}

/**
 Returns the managed object model for the application.
 */
- (NSManagedObjectModel *)managedObjectModel
{
    if (_managedObjectModel != nil)
    {
        return _managedObjectModel;
    }
    NSURL *modelURL = [[NSBundle mainBundle] URLForResource:__managedObjectModelName withExtension:@"momd"];
    if (!modelURL) modelURL = [[NSBundle mainBundle] URLForResource:__managedObjectModelName withExtension:@"mom"];
    _managedObjectModel = [[NSManagedObjectModel alloc] initWithContentsOfURL:modelURL];
    return _managedObjectModel;
}

/**
 Returns the persistent store coordinator for the application.
 */
- (NSPersistentStoreCoordinator *)persistentStoreCoordinator
{
    if (_persistentStoreCoordinator != nil)
    {
        return _persistentStoreCoordinator;
    }
    
    NSString *dbName = __databaseName;
    
#ifdef TEST
    dbName = __databaseNameTest;
#endif
    
    NSURL *storeURL = [[self applicationDocumentsDirectory] URLByAppendingPathComponent:dbName];
    
    NSError *error = nil;
    NSDictionary *options = [NSDictionary dictionaryWithObjectsAndKeys:
                             [NSNumber numberWithBool:YES], NSMigratePersistentStoresAutomaticallyOption,
                             [NSNumber numberWithBool:YES], NSInferMappingModelAutomaticallyOption, nil];
    _persistentStoreCoordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:[self managedObjectModel]];
    if (![_persistentStoreCoordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:nil URL:storeURL options:options error:&error])
    {
        
        NSLog(@"Unresolved error %@, %@\n\n", error, [error userInfo]);
        //abort();
    }
    
    return _persistentStoreCoordinator;
}

#pragma mark - Application's Documents directory

/**
 Returns the URL to the application's Documents directory.
 */
- (NSURL *)applicationDocumentsDirectory
{
    return [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
}

- (void)saveContext
{
    NSError *error = nil;
    NSManagedObjectContext *aManagedObjectContext = self.managedObjectContext;
    if (aManagedObjectContext != nil)
    {
        BOOL isChanged = [aManagedObjectContext hasChanges];
        if (isChanged && ![aManagedObjectContext save:&error])
        {
            NSLog(@"Unresolved error %@, %@\n\n", error, [error userInfo]);
            //abort();
        }
    }
}

@end

File: CoreData/ZFRCoreData.swift
//
//  ZFRCoreData.swift
//  Pods
//
//  Created by Francesco Frison on 6/11/14.
//
//

import Foundation
import CoreData


struct ZFRCoreDataConfig {
    var mocName : String
    var dbName : String
    
    init (mocName : String) {
        self.mocName = mocName
        self.dbName = ZFRCoreDataConfig.dbNameFromMocName(mocName)
    }
    
    init() {
        self.mocName = NSBundle.mainBundle().objectForInfoDictionaryKey("CFBundleDisplayName") as String
        self.dbName = ZFRCoreDataConfig.dbNameFromMocName(self.mocName)
    }
    
    static func dbNameFromMocName(mocName: String) -> String {
        var dbName : String
        #if DEBUG
            dbName = "\(mocName).sqlite"
            #else
            dbName = "test-\(mocName).sqlite"
        #endif
        return dbName
    }

}

class ZFRCoreData {
    
    
    var config : ZFRCoreDataConfig {
    get {
        return ZFRCoreDataConfig()
    }
    }
    
    var managedObjectModel : NSManagedObjectModel? {
    get {
        
        
        var url = NSBundle().URLForResource(self.config.mocName, withExtension: "momd")
        if !url {
            url = NSBundle().URLForResource(self.config.mocName, withExtension: "mom")
        }
        
        return NSManagedObjectModel(contentsOfURL: url)
    }
    }
    
    var persistentStoreCoordinator :NSPersistentStoreCoordinator? {
    get {
        let storeC : NSPersistentStoreCoordinator? = NSPersistentStoreCoordinator(managedObjectModel: self.managedObjectModel)
        
        if !storeC {
            return nil
        }
        
        let storeURL = NSBundle.mainBundle().resourceURL.URLByAppendingPathComponent(self.config.dbName)
        
        var options : Dictionary<String, Bool> = Dictionary()
        options[NSMigratePersistentStoresAutomaticallyOption] = true
        options[NSInferMappingModelAutomaticallyOption] = true
        
        storeC!.addPersistentStoreWithType(NSSQLiteStoreType, configuration: nil, URL: storeURL, options: options, error:nil)
        return storeC
    }
    }
    var managedObjectContext: NSManagedObjectContext? {
    get {
        let moc : NSManagedObjectContext? = NSManagedObjectContext()
        if !moc {
            return nil
        }
        moc!.persistentStoreCoordinator = self.persistentStoreCoordinator
        return moc
    }
    }
    

}
